pygen/pygen_src/isa/riscv_cov_instr.py

```diff
@@ -0,0 +1,530 @@
+            return self.rs1_value.get_val() != self.rs2_value.get_val()
+        elif self.instr.name == "C_BNEZ":
+            return self.rs1_value.get_val() != 0
+        elif self.instr.name == "BLT" or self.instr.name == "BLTU":
+            return self.rs1_value.get_val() < self.rs2_value.get_val()
+        elif self.instr.name == "BGE" or self.instr.name == "BGEU":
+            return self.rs1_value.get_val() >= self.rs2_value.get_val()
+        else:
+            logging.error("Unexpected instruction {}".format(self.instr.name))
+
+    def get_logical_similarity(self):
+        val1 = vsc.int_t(rcs.XLEN, self.rs1_value.get_val())
+        val2 = vsc.int_t(rcs.XLEN, (self.imm.get_val() if
+                                    self.format == riscv_instr_format_t.I_FORMAT
+                                    else self.rs2_value.val))
+        temp = bin(val1.get_val() ^ val2.get_val())
+        bit_difference = len([[ones for ones in temp[2:] if ones == '1']])
+        if val1.get_val() == val2.get_val():
+            return logical_similarity_e["IDENTICAL"]
+        elif bit_difference == 32:
+            return logical_similarity_e["OPPOSITE"]
+        elif bit_difference < 5:
+            return logical_similarity_e["SIMILAR"]
+        else:
+            return logical_similarity_e["DIFFERENT"]
+
+    def check_hazard_condition(self, pre_instr):
+        if pre_instr.has_rd:
+            if ((self.has_rs1 and self.rs1 == pre_instr.rd) or
+                    (self.has_rs2 and self.rs1 == pre_instr.rd)):
+                self.gpr_hazard = hazard_e["RAW_HAZARD"]
+            elif self.has_rd and self.rd == pre_instr.rd:
+                self.gpr_hazard = hazard_e["WAW_HAZARD"]
+            elif (self.has_rd and
+                  ((pre_instr.has_rs1 and (pre_instr.rs1 == self.rd)) or
+                   (pre_instr.has_rs2 and (pre_instr.rs2 == self.rd)))):
+                self.gpr_hazard = hazard_e["WAR_HAZARD"]
+            else:
+                self.gpr_hazard = hazard_e["NO_HAZARD"]
+        if self.category == riscv_instr_category_t.LOAD:
+            if (pre_instr.category == riscv_instr_category_t.STORE and
+                    pre_instr.mem_addr.get_val() == self.mem_addr.get_val()):
+                self.lsu_hazard = hazard_e["RAW_HAZARD"]
+            else:
+                self.lsu_hazard = hazard_e["NO_HAZARD"]
+        if self.category == riscv_instr_category_t.STORE:
+            if (pre_instr.category == riscv_instr_category_t.STORE and
+                    pre_instr.mem_addr.get_val() == self.mem_addr.get_val()):
+                self.lsu_hazard = hazard_e["WAW_HAZARD"]
+            elif (pre_instr.category == riscv_instr_category_t.LOAD and
+                  pre_instr.mem_addr.get_val() == self.mem_addr.get_val()):
+                self.lsu_hazard = hazard_e["WAR_HAZARD"]
+            else:
+                self.lsu_hazard = hazard_e["NO_HAZARD"]
+        logging.info("Pre: {}, Cur: {}, Hazard: {}/{}".format(
```

Comment:
I'd suggest making this a `logging.debug()` call instead

Reply:
Sure, thanks Udi.

Reply:
I changed it it to a debug message, it prints prev/current PCs, instruction names, and also type of the hazards. Adding related registers/their values will depend on the hazard type (which would make the code a bit ugly and large), so I didn't include them in the debug message (I think having PC address as well as instruction names/hazard type will be enough, right?)

Reply:
Yeah, I think that is fine - I think with the provided information it should be straightforward enough to refer to the CSV file and find the two instructions that cause the hazard.